workflows/general/bam-to-fastqs.wdl

```diff
@@ -36,50 +36,47 @@ version 1.0
 
 
 import "../../tools/samtools.wdl"
-import "../../tools/picard.wdl"
 import "../../tools/fq.wdl"
 
 workflow bam_to_fastqs {
     input {
         File bam
-        Boolean paired = true
+        Boolean paired_end = true
         Boolean use_all_cores = false
         Int? max_retries
     }
 
     parameter_meta {
         bam: "BAM file to split into FastQs"
-        paired: "Is the data paired-end (true) or single-end (false)?"
+        paired_end: "Is the data paired-end (true) or single-end (false)?"
         use_all_cores: "Use all cores for multi-core steps?"
         max_retries: "Number of times to retry failed steps. Overrides task level defaults."
     }
 
     call samtools.quickcheck { input: bam=bam, max_retries=max_retries }
     call samtools.split { input: bam=bam, use_all_cores=use_all_cores, max_retries=max_retries }
     scatter (split_bam in split.split_bams) {
-        call picard.bam_to_fastq { input: bam=split_bam, paired=paired, max_retries=max_retries }
-    }
-
-    if (paired) {
-        scatter (reads in zip(bam_to_fastq.read1, bam_to_fastq.read2)) {
-            call fq.fqlint as fqlint_pair { input:
-                read1=reads.left,
-                read2=reads.right,
-                max_retries=max_retries
-            }
+        call samtools.collate_to_fastq as bam_to_fastq { input:
+            bam=split_bam,
+            paired_end=paired_end,
+            interleaved=false,  # matches default but prevents user from overriding
+            use_all_cores=use_all_cores,
+            max_retries=max_retries
         }
     }
-    if (! paired) {
-        scatter (reads in bam_to_fastq.read1) {
-            call fq.fqlint as fqlint_single { input:
-                read1=reads,
-                max_retries=max_retries
-            }
+
+    scatter (reads in 
+        zip(bam_to_fastq.read_one_fastq_gz, bam_to_fastq.read_two_fastq_gz)
+    ) {
+        call fq.fqlint { input:
+            read1=select_first([reads.left, "undefined"]),
+            read2=reads.right,
```

Comment:
Is this correct? I would expect read one to always exist and read2 to be the optional input.

Reply:
Yes this is an unfortunate side effect of enabling the full spectrum of options in `samtools fastq`. The `_R1` file won't be generated if `interleaved==true` (because then no `_R*` extension is used). However we know from the context of this workflow that the `_R1` file will always exist. Just WDL doesn't know that. It is impossible (or should be impossible) to trip the `"undefined"` part.

The `read2` parameter accepts `File?` input so no special handling of it needs to be made.